src/serial.rs

```diff
@@ -584,15 +585,21 @@ macro_rules! serialdma {
                 }
             }
 
-            impl<B> crate::dma::CircReadDma<B, u8> for $rxdma where B: as_slice::AsMutSlice<Element=u8> {
-                fn circ_read(mut self, buffer: &'static mut [B; 2],
+            impl<B> crate::dma::CircReadDma<B, u8> for $rxdma
+            where
+                &'static mut [B; 2]: StaticWriteBuffer<Word = u8>,
+                B: 'static,
+            {
+                fn circ_read(mut self, mut buffer: &'static mut [B; 2],
                 ) -> CircBuffer<B, Self>
                 {
                     {
-                        let buffer = buffer[0].as_mut_slice();
+                        // NOTE(unsafe) We own the buffer now and we won't call other `&mut` on it
+                        // until the end of the transfer.
+                        let (ptr, len) = unsafe { buffer.static_write_buffer() };
                         self.channel.set_peripheral_address(unsafe{ &(*$USARTX::ptr()).dr as *const _ as u32 }, false);
-                        self.channel.set_memory_address(buffer.as_ptr() as u32, true);
-                        self.channel.set_transfer_length(buffer.len() * 2);
+                        self.channel.set_memory_address(ptr as u32, true);
+                        self.channel.set_transfer_length(len);
 
                         atomic::compiler_fence(Ordering::Release);
 
```

Comment:
Does opening a new scope for this still serve a purpose?

Reply:
It limits the `ptr` scope, but not really necessary, it was there before so I didn't touch it.

Reply:
Yea, that's what I figured. I'd say it's better to remove the scope then, just to avoid people having to ask that question every time they look at the code :D